core_lib/src/interface/editor.cpp

```diff
@@ -1077,86 +1058,60 @@ KeyFrame* Editor::addNewKey()
 KeyFrame* Editor::addKeyFrame(int layerNumber, int frameIndex)
 {
     Layer* layer = mObject->getLayer(layerNumber);
-    Q_ASSERT(layer);
-
     if (!layer->visible())
     {
         mScribbleArea->showLayerNotVisibleWarning();
         return nullptr;
     }
-
-    // Find next available space for a keyframe (where either no key exists or there is an empty sound key)
-    while (layer->keyExists(frameIndex))
-    {
-        if (layer->type() == Layer::SOUND
-            && layer->getKeyFrameAt(frameIndex)->fileName().isEmpty()
-            && layer->removeKeyFrame(frameIndex))
-        {
-            break;
-        }
-        else
-        {
-            frameIndex += 1;
-        }
-    }
-
-    bool ok = layer->addNewKeyFrameAt(frameIndex);
-    if (ok)
-    {
-        scrubTo(frameIndex); // currentFrameChanged() emit inside.
-        emit frameModified(frameIndex);
-        layers()->notifyAnimationLengthChanged();
-    }
-    return layer->getKeyFrameAt(frameIndex);
+    return addKeyFrameToLayer(layer, frameIndex, false);
 }
 
 void Editor::removeKeyAtLayerId(int layerId, int frameIndex)
 {
     Layer* layer = layers()->findLayerById(layerId);
-    int layerIndex = layers()->getLayerIndex(layer);
+    removeKeyAtLayer(layer, frameIndex);
+}
+
+void Editor::removeKeyAtLayer(Layer *layer, int frameIndex, bool shouldBackup)
+{
+    if (layer == nullptr)
+    {
+        Q_ASSERT(false);
+        return;
+    }
+
     if (!layer->keyExistsWhichCovers(frameIndex))
     {
         return;
     }
 
+    if (shouldBackup)
+    {
+        backup(tr("Remove frame"));
+    }
```

Comment:
When would you ever do a state change and not require a backup of the action? we won't be able to go back otherwise.

Reply:
removeKeyAtLayerId wasn't backing it up before these changes. You want to remove a key and not do a state change when you're calling that from backupelements so you don't get into some infinite backup loop. Key removals triggered by an action use removeKey() which is still does a back.

Reply:
Agh.. you're right, I forgot about that one. I couldn't remember it well enough and the diff didn't make it obvious.